text/0957-modernized-scheduler.md

```diff
@@ -0,0 +1,646 @@
+hitting the dreaded "backtracking render" error when it should have been safe to have read and
+written multiple times.
+
+3) Awaiting a data fetch and assuming that the work done after await occurs before render, when
+due to Ember's current zealous flushing it is taking place post-render.
+
+
+## Detailed design
+
+The scheduler interface defines several "phases" of when work should be done
+that align to concepts in the browser's render cycle. The primary goal is
+to reduce the number of times Ember needs to alter the DOM per frame to 0 or 1
+as often as possible.
+
+We also want to help applications coordinate work in a way that avoids the pitfalls
+of DOM read/write interleaving. This means we need to provide both a mechanism for
+applications to schedule work that occurs *after* the render but before the browser
+has painted, and a mechanism by which to schedule work *after* that work but also
+before the browser has painted.
+
+To understand the phases of work to be done, it helps to understand a bit about
+the mechanics of the browser's render cycle.
+
+### Frames
+
+The scheduler interface conceptualizes work as belonging to a "Frame", where a Frame
+constitutes the time between when states of the DOM are observable to a user.
+
+Below is a rough approximation of how browsers schedule various work.
+
+![Frame Overview](../images/0957-frame-overview.svg)
+
+Tasks are things such as `setTimout`, callbacks for `DOMEvent`s, or the completion
+of an `xhr` or `fetch` request. The browser will execute as many tasks as it can
+until it has met or passed the point at which it would like to try to update the screen.
+
+`Microtasks` are callbacks executed via  `createMicrotask(callback)`, promise completion
+such as `Promise.resolve().then(callback)`, `MutationObserver` callbacks, or `MessageChannel`
+callbacks. Microtasks can recursively schedule new microtasks, and the browser will not
+move on to another task or continue towards updating the screen until the queue is exhausted.
+
+`FrameTasks` are callbacks executed for `requestAnimationFrame(callback)`. The browser
+will execute all `FrameTasks` that existed when it began flushing the queue, but callbacks
+may not schedule new `FrameTasks` recursively. E.g. any new calls to `requestAnimationFrame`
+will go into the "next" Frame.
+
+Similar to `FrameTasks`, `ResizeTasks` are for `ResizeObserver` callbacks and `IntersectionTasks`
+are for `IntersectionObserver` callbacks. These queues can recursively schedule, though if
+too much recursion is encountered by a `ResizeObserver` is errors and tries again on the next frame.
+
+### Phases
+
+With this understanding of the lifecycle of a Frame, roughly speaking we start to see a few
+patterns for how we might want to schedule work in a way that seamlessly coordinates with the
+framework and other parts of our application.
+
+We divide this work into 6 conceptual phases:
+
+- `tasks` - work that updates reactive state
```

Comment:
we may need a sync phase between `tasks` and `render`

this phase would be before we try rendering.

e.g.:
- sync runs before render
  - and after component construction
  - not in the render tracking frame / does not invalidate the render tracking frame
- must not run when the parent context is torn down

our purpose for on.sync is to create a mechanism for constructing a stateful object with a lifetime ("resource") that can use reactive data as part of its constructor. when the stateful data changes, the resource is cleaned up and reconstructed. also, when the parent lifetime is destroyed, the resource is cleaned up. and then finally, the resource object itself has a stable identity regardless of whether it has the internal resource has been synchronized or reset.

tl;dr: stable object on the outside, unstable object on the inside (which gets reset based on reactivity)